Support generic os.PathLike instead of pathlib.Path

olebole opened the pull request:
### Description

This PR replaces all appearances of 

```Python
 if isinstance(fileobj, pathlib.Path): 
     fileobj = str(fileobj) 
```
with 
```Python
if isinstance(fileobs, os.PathLike):
    fileobj = fileobj.__fspath__()
```
Rationale: In Python, paths are not necessarily strings or `pathlib.Path` objects, but anything that is `os.Pathlike` is acceptable in `open()`. One use case here is a specific path like object that retrieves/builds a file just before it is opened (f.e. for a cache).
The canonical way to access the file path on the OS is then done by `__fspath__()`, not by `str()`.

Fixes #11579

Cadair commented:
direct use of `__fspath__` feels odd. What's the difference between that and `str()`?

olebole commented:
@Cadair There is no guarantee that `str()` actually returns the path in the file system. This is true for `pathlib.Path`, but other `os.PathLike` classes may not even implement a proper `__str__()` method to return the path. For example, the `DummyPath` class in #11579 is a valid implementation of `os.PathLike`, but would fail when using `str()`.

In contrast, `__fspath__()` is a [well-defined interface](https://docs.python.org/3/library/os.html?highlight=pathlike#os.PathLike.__fspath__), so I don't see why one shouldn't use it.

Cadair commented:
That makes sense, but seems odd that class doesn't have a non-magic method for doing that.

adrn commented:
Nice! I think this is a great idea, and is almost a bug fix. Could you add a test, even taken from your original issue?

pllim commented on docs/changes/io.fits/11580.feature.rst:
```diff
@@ -0,0 +1 @@
+Enable the use of ``os.PathLike´´ objects when dealing with (mainly FITS) files.
```
This would fix RTD build.
```suggestion
Enable the use of ``os.PathLike`` objects when dealing with (mainly FITS) files.
```

pllim commented:
Thanks! Agree with @adrn that new tests are needed to cover the new changes.

pllim commented:
Also perhaps add an example in user-facing doc? Otherwise, this new improvement is hard to discover.

saimn commented:
> In contrast, __fspath__() is a well-defined interface, so I don't see why one shouldn't use it.

Indeed, but then you should use `os.fspath` instead:
https://docs.python.org/3/whatsnew/3.6.html#pep-519-adding-a-file-system-path-protocol
https://docs.python.org/3/glossary.html#term-path-like-object

olebole commented:
Thank you for the suggestions! I now fixed the changelog RST, and replaced `__fspath__()` with `os.fspath()`. I will add a simple test later.
Concerning an example in user-faced doc: I don't have a good, simple example that could serve here. My own use case is a Gitlab file cache (to overcome the problem that one cannot easily selectively mount some files from a Git repository via submodule), but I have no 10-lines example. Suggestions would be welcome here.

pep8speaks commented:
Hello @olebole :wave:! It looks like you've made some changes in your pull request, so I've checked the code again for style.










* In the file [`astropy/io/fits/tests/test_core.py`](https://github.com/astropy/astropy/blob/f6b5f24f4e24ff74d12fcc6eb756d0b60e27f1d5/astropy/io/fits/tests/test_core.py):

> [Line 1391:1](https://github.com/astropy/astropy/blob/f6b5f24f4e24ff74d12fcc6eb756d0b60e27f1d5/astropy/io/fits/tests/test_core.py#L1391): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

mhvk reviewed (approved):
Maybe best for @cadair to do a proper review, but this looks all OK to me. Nice!

saimn reviewed (approved):
Looks good to me, thanks @olebole !

mhvk commented:
Two approvals would seem good enough. Let's get it in. Thanks, @olebole!

nstarman commented:
Oh no! Lots of stuff in the docs should now by "path-like", see #11118.

nstarman commented on astropy/io/fits/util.py:
```diff
@@ -26,8 +25,7 @@
 
-from os import PathLike
-path_like = (str, PathLike)
+path_like = (str, os.PathLike)
```
according to https://docs.python.org/3/glossary.html#term-path-like-object, path-like should include bytes

nstarman replied:
The three components of python's path-like are string, byte, os.PathLike

nstarman commented on astropy/io/registry.py:
```diff
@@ -24,7 +23,7 @@
 _identifiers = OrderedDict()
 
-PATH_TYPES = (str, pathlib.Path)
+PATH_TYPES = (str, os.PathLike)
```
same. should include `bytes`

nstarman commented:
I've added this as a TODO in the above issue
